Haskell/Haskell.sublime-syntax

```diff
@@ -5,68 +5,140 @@ name: Haskell
 file_extensions:
   - hs
 scope: source.haskell
+
+variables:
+  # In case this regex seems overly general,
+  # note that Haskell permits the definition of new operators
+  # which can be nearly any string of punctuation characters,
+  # such as $%^&*.
+  operator_infix: '[*|!%$?~+:\-.=</>\\]+'
+  operator_parens: '(?:{{operator_infix}}|,+)'
+  module_name: (?:[A-Z][A-Za-z._']*)
```

Comment:
Unicode should probably be considered here as well.

Reply:
How would I differentiate a unicode name from a unicode operator?

Reply:
> How would I differentiate a unicode name and a unicode operator?

Hmm, I'll see if I can find a more comprehensive description of Haskell's lexical syntax in this area. It's entirely possible that you can't. In the Scala mode, for example, I gave up entirely on this and just matched identifiers as… identifiers, since there's no meaningful distinction. Haskell does distinguish both at definition site and at call site, but only in a context-sensitive fashion:

- At definition site, parens/backticks may distinguish names from operators, as well as infix usage of the type-declared name (this is the context-sensitive bit)
- At call site, parens/backticks may also be found, but also the use of an identifier in infix position rather than prefix, which should theoretically be parseable

There's got to be *some* distinguishing lexical feature though, which should give us something to hang our hats on. I'll do some digging